### DISCUSSION TITLE
`special` vs built-in schema + `custom` validation?

### DISCUSSION THREAD

**Discussion by jindong-zhannng** [open-circle/valibot#555] [General]
Hi folks,

I just found basically both of these 2 ways can do the same thing. e.g.

**`special`**

```ts
const NumericStringSchema = v.special<string>((i) => typeof i === 'string' && !isNaN(Number(i)))
```

**built-in schema + `custom`**

```ts
const NumericStringSchema = v.string([v.custom((i) => !isNaN(Number(i)))])
```

Which way is more preferred?

**Comment by fabian-hiller** [comment]
I recommend using `string` with `custom`. `special` is for more special or custom data types. What exactly are you trying to validate? Maybe it makes sense to add a new action validation so you can write the following:

```ts
const NumericStringSchema = v.string([v.numeric()]);
```

**Reply by jindong-zhannng** [reply]
My use case is parsing string-based form inputs to typed data.
For example if I have a input box that should only allow number inputs. When I use `transform` api to do the parsing, I'd like to validate it with `v.string([v.numeric()])` firstly, then transform string inputs to numbers. So that I can gain meaningful error messages from the validation if input is invalid.

> Maybe it makes sense to add a new action validation

Yes a `numeric` validation for strings may makes sense if other people like it also. 
But is there an api I can use to define my own validation actions? The `custom` api? As I mentioned above, in my use case I may need more various validations to match different input types. I hope I can get more meaningful errors when validation fails. e.g. by overriding `context` or `expected` properties, inserting received value into error message. However the `custom` api seems unable to do these now.

**Reply by fabian-hiller** [reply]
> Yes a `numeric` validation for strings may makes sense if other people like it also.

I like this idea and think we should add it. Feel free to create an issue or PR that merges it into the branch of #502.

> I hope I can get more meaningful errors when validation fails. e.g. by overriding `context` or `expected` properties

This will require you to write your own validation action. But it is quite easy, you can just copy the source code of a similar validation action from our source code and change it to your needs.
